lib/global-styles.php

```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
+ * @return array Global Styles tree.
  */
-function gutenberg_enqueue_global_styles_assets() {
-	if ( ! locate_template( 'experimental-theme.json' ) ) {
-		return;
-	}
-	$default_global_styles_path = dirname( dirname( __FILE__ ) ) . '/experimental-default-global-styles.json';
-	$default_global_styles      = null;
-
-	if ( file_exists( $default_global_styles_path ) ) {
-		$default_global_styles = json_decode(
-			file_get_contents( $default_global_styles_path ),
+function gutenberg_experimental_global_styles_get_from_file( $global_styles_path ) {
+	$global_styles = array();
+	if ( file_exists( $global_styles_path ) ) {
+		$decoded_file = json_decode(
+			file_get_contents( $global_styles_path ),
 			true
 		);
+		if ( is_array( $decoded_file ) ) {
+			$global_styles = $decoded_file;
+		}
 	}
+	return $global_styles;
+}
+
+/**
+ * Returns an array containing the user's Global Styles
+ * or a void array if none.
+ *
+ * @return array Global Styles tree.
+ */
+function gutenberg_experimental_global_styles_get_user() {
+	$global_styles = array();
+	$user_cpt      = gutenberg_experimental_global_styles_get_user_cpt( array( 'publish' ) );
+	if ( array_key_exists( 'post_content', $user_cpt ) ) {
+		$decoded_data = json_decode( $user_cpt['post_content'], true );
+		if ( is_array( $decoded_data ) ) {
+			$global_styles = $decoded_data;
+		}
+	}
+
+	return $global_styles;
+}
+
+/**
+ * Returns the CPT that contains the user's Global Styles
+ * for the current theme or a void array if none found.
+ * It can also create and return a new draft CPT.
+ *
+ * @param array $post_status_filter Filter CPT by post status.
+ *                                  By default, only fetches published posts.
+ * @param bool  $should_create_draft Whether a new draft should be created
+ *                                   if no CPT was found. False by default.
+ * @return array Custom Post Type for the user's Global Styles.
+ */
+function gutenberg_experimental_global_styles_get_user_cpt( $post_status_filter = array( 'publish' ), $should_create_draft = false ) {
+	$user_cpt         = array();
+	$post_type_filter = 'wp_global_styles';
+	$post_name_filter = 'wp-global-styles-' . strtolower( wp_get_theme()->get( 'TextDomain' ) );
+	$recent_posts     = wp_get_recent_posts(
+		array(
+			'numberposts' => 1,
+			'orderby'     => 'date',
+			'order'       => 'desc',
+			'post_type'   => $post_type_filter,
+			'post_status' => $post_status_filter,
+			'name'        => $post_name_filter,
+		)
+	);
 
-	$theme_json_path          = locate_template( 'experimental-theme.json' );
-	$theme_json_global_styles = null;
-	if ( $theme_json_path ) {
-		$theme_json_global_styles = json_decode(
-			file_get_contents( $theme_json_path ),
+	if ( is_array( $recent_posts ) && ( count( $recent_posts ) === 1 ) ) {
+		$user_cpt = $recent_posts[0];
+	} elseif ( $should_create_draft ) {
+		$cpt_post_id = wp_insert_post(
+			array(
+				'post_content' => '{}',
+				'post_status'  => 'draft',
```

Comment:
I think we may have a bug, in gutenberg_experimental_global_styles_get_user we are querying published posts, but we insert a draft post. On the next execution, a published post will still not exist and we will probably create another one.
I think we may be able to create the post with the published status ( the empty object will not do anything);

Reply:
Note the optional parameter `$should_create_draft`, which is false by default. We only insert a draft if there is none in the editor call `gutenberg_experimental_global_styles_get_user_cpt_id`. That's the only place we need a pre-existing CPT: the entities API only works with pre-existing objects, can't create new ones from the client-side.